### ISSUE TITLE
Android - coreclr_initialize failed - status: 0x80004005 because of mkfifo call

### ISSUE THREAD

**Issue by qmfrederik** [dotnet/coreclr#9780]
When cross-building CoreCLR for Android, running a simple Hello World application fails with the following output:

```
coreclr_initialize failed - status: 0x80004005
```

Setting `COMPlus_BreakOnHR=80004005` and some stepping through the code with lldb reveals that the root cause for this failure is here:

```
* thread #1, name = 'corerun', stop reason = breakpoint 10.1
    frame #0: libcoreclr.so`TwoWayPipe::CreateServer(this=0x000000555ee45938, id=13506) at twowaypipe.cpp:29
   26
   27       unlink(m_inPipeName);
   28
-> 29       if (mkfifo(m_inPipeName, S_IRWXU) == -1)
   30       {
   31           return false;
   32       }
```

with `m_inPipeName` set to `/data/local/tmp/clr-debug-pipe-13506-424747-in`.

This is the same issue as seen previously seen by @cydhaselton and @janvorli in https://github.com/dotnet/coreclr/issues/917#issuecomment-262652992.

 Long story short (see e.g. [this discussion on the android-ndk mailing list](https://groups.google.com/forum/#!topic/android-ndk/lD-V7Nxe5y4)), there is no reliable way to create a named pipe on Android that can be shared accross processes/applications. (@cydhaselton was able to do it in Termux because Termux is deployed as an Android application).

Looking at the code, it appears these named pipes are used to enable a remote debugger to connect. I'm not sure that scenario makes a lot of sense on Android, and it appears this code is guarded with an `FEATURE_DBGIPC_TRANSPORT_DI` feature flag.

Net, I'd like to understand if it would be safe to disable this feature alltogether on Android? Is there a specific location in the CMake files where we could control this flag?

**Comment by qmfrederik** [comment]
@janvorli I guess this one is mainly for you - can we disable the debug transport over named pipes in CoreCLR for the Android build? Would that require changes in the CMake files?

**Comment by janvorli** [comment]
@qmfrederik I don't know. You can try to disable that feature in clrdefinitions.cmake and see what happens. 
But @mikem8361 should be able to give you the right answer.

**Comment by cydhaselton** [comment]
@qmfrederik @janvorli and @mikem8361 If this can be disabled regardless of whether or not coreclr is run within a hosted application, like Termux, that would be best.  I'm fairly certain that named pipes were never truly functional when I was working within Termux.

**Comment by cydhaselton** [comment]
Out of curiosity, is named pipe functionality included in the Release build? If not, perhaps a conditional could be added to exclude from Android build.

**Comment by cydhaselton** [comment]
@qmfrederik Cherry-pick[ android-disable-debug](https://github.com/cydhaselton/coreclr/tree/android-disable-debug) from my fork…see if that resolves the issue

**Comment by cydhaselton** [comment]
@janvorli Is there a reason for the case in the [following lines](https://github.com/dotnet/coreclr/blob/32f0f9721afb584b4a14d69135bea7ddc129f755/src/debug/di/platformspecific.cpp#L20-L23) in coreclr/src/debug/di/platformspecific.cpp?

**Comment by qmfrederik** [comment]
FYI - with #9809 I got the CLR to load and it starts executing my "Hello World" assembly, at which point an exception occurrs and unwinding the stack fails:

```
fcarlier@ubuntu:~/git/coreclr$ adb shell LD_LIBRARY_PATH=/data/local/tmp/coredroid/ /data/loc                                                                            al/tmp/coredroid/corerun /data/local/tmp/coredroid/helloworld.dll

Assert failure(PID 13990 [0x000036a6], Thread: 13990 [0x36a6]): !"Thread::VirtualUnwindToFirs                                                                            tManagedCallFrame: PAL_VirtualUnwind failed"
    File: /home/fcarlier/git/coreclr/src/vm/stackwalk.cpp Line: 770
    Image: /data/local/tmp/coredroid/corerun

Aborted
```

and here's the full backtrace:

```
  * frame #0: 0x0000007fb7f54e5c libc.so`tgkill + 8
    frame #1: 0x0000007fb7f12f20 libc.so`pthread_kill + 164
    frame #2: 0x0000007fb7f14454 libc.so`raise + 32
    frame #3: 0x0000007fb7f0df00 libc.so`abort + 64
    frame #4: libcoreclr.so`PROCEndProcess(hProcess=0x00000000ffffff01, uExitCode=123456789, bTerminateUnconditionally=YES) at process.cpp:1385
    frame #5: libcoreclr.so`::TerminateProcess(hProcess=0x00000000ffffff01, uExitCode=123456789) at process.cpp:1301
    frame #6: libcoreclr.so`TerminateOnAssert() at debug.cpp:183
    frame #7: libcoreclr.so`::_DbgBreakCheck(szFile=0x0000007fb7adbd2c, iLine=770, szExpr=0x0000007fb7adbdcd, fConstrained=NO) at debug.cpp:436
    frame #8: libcoreclr.so`_DbgBreakCheckNoThrow(szFile=0x0000007fb7adbd2c, iLine=770, szExpr=0x0000007fb7adbdcd, fConstrained=NO) at debug.cpp:548
    frame #9: libcoreclr.so`::DbgAssertDialog(szFile=0x0000007fb7adbd2c, iLine=770, szExpr=0x0000007fb7adbdcd) at debug.cpp:735
    frame #10: libcoreclr.so`Thread::VirtualUnwindToFirstManagedCallFrame(pContext=0x0000007fffffe410) at stackwalk.cpp:770
    frame #11: libcoreclr.so`DispatchManagedException(ex=0x0000007fffffe9f8, isHardwareException=false) at exceptionhandling.cpp:4632
    frame #12: libcoreclr.so`::PreStubWorker(pTransitionBlock=0x0000007fffffeab0, pMD=0x0000007f3d5219c0) at prestub.cpp:1076
    frame #13: 0x0000007fb7526790 libcoreclr.so`ThePreStub + 80
    frame #14: 0x0000007f3d93a1f4
    frame #15: 0x0000007fb7527254 libcoreclr.so`CallDescrWorkerInternal + 132
    frame #16: libcoreclr.so`CallDescrWorkerWithHandler(pCallDescrData=0x0000007fffffef08, fCriticalCall=NO) at callhelpers.cpp:78
    frame #17: libcoreclr.so`MethodDescCallSite::CallTargetWorker(this=0x0000007ffffff088, pArguments=0x0000007ffffff018, pReturnValue=0x0000000000000000, cbReturnValue=0) at callhelpers.cpp:645
    frame #18: libcoreclr.so`MethodDescCallSite::Call(this=0x0000007ffffff088, pArguments=0x0000007ffffff018) at callhelpers.h:433
    frame #19: libcoreclr.so`RunMain(this=0x0000007ffffff1d8, pParam=0x0000007ffffff2e0)::$_1::operator()(RunMain(MethodDesc*, short, int*, REF<PtrArray>*) const::Param*) const::'lambda'(RunMain(MethodDesc*, short, int*, REF<PtrArray>*) const::Param*)::operator()(RunMain(MethodDesc*, short, int*, REF<PtrArray>*) const::Param*) const at assembly.cpp:2057
    frame #20: libcoreclr.so`RunMain(this=0x0000007ffffff2d0, __EXparam=0x0000007ffffff2e0)::$_1::operator()(RunMain(MethodDesc*, short, int*, REF<PtrArray>*)::Param*) const at assembly.cpp:2075
    frame #21: libcoreclr.so`RunMain(pFD=0x0000007f3d51f128, numSkipArgs=1, piRetVal=0x0000007ffffff3fc, stringArgs=0x0000007ffffff850) at assembly.cpp:2075
    frame #22: libcoreclr.so`Assembly::ExecuteMainMethod(this=0x000000555b647840, stringArgs=0x0000007ffffff850, waitForOtherThreads=YES) at assembly.cpp:2157
    frame #23: libcoreclr.so`CorHost2::ExecuteAssembly(this=0x000000555b5c6ec0, dwAppDomainId=1, pwzAssemblyPath=u"/data/local/tmp/coredroid/helloworld.dll", argc=0, argv=0x0000000000000000, pReturnValue=0x0000007ffffffb74) at corhost.cpp:501
    frame #24: libcoreclr.so`::coreclr_execute_assembly(hostHandle=0x000000555b5c6ec0, domainId=1, argc=0, argv=0x0000000000000000, managedAssemblyPath=0x000000555b5bc0a8, exitCode=0x0000007ffffffb74) at unixinterface.cpp:376
    frame #25: corerun`ExecuteManagedAssembly(currentExeAbsolutePath=0x000000555b5bc058, clrFilesAbsolutePath=0x000000555b5bc0f8, managedAssemblyAbsolutePath=0x000000555b5bc0a8, managedAssemblyArgc=0, managedAssemblyArgv=0x0000000000000000) at coreruncommon.cpp:420
    frame #26: corerun`corerun(argc=2, argv=0x0000007ffffffdd8) at corerun.cpp:149
    frame #27: corerun`main(argc=2, argv=0x0000007ffffffdd8) at corerun.cpp:161
    frame #28: 0x0000007fb7f0c8ec libc.so`__libc_init + 104
    frame #29: 0x000000555b570304 corerun`do_arm64_start + 64
    frame #30: 0x0000007fb7fdf650 linker64`__dl__start + 8
``` 

So I guess there are two separate issues - one that causes the code to fail on Android, and one that cause the unwinding to fail.

I'll try getting the LLDB plugin to work to see what the managed stack trace looks like; that'll probably be something for later this week.

**Comment by cydhaselton** [comment]
@janvorli Did you see my earlier question?

**Comment by janvorli** [comment]
@cydhaselton I have seen it, but then couldn't find it again. There is no reason for the casing. On Windows, the filesystem is not case sensitive and so it doesn't matter if the case in the #include doesn't match the actual case of the file. In the early days of porting coreclr to Unix, we have changed all files that were included / compiled on Linux to lowercase to fix the problem. But some files that were windows specific remained.

**Comment by cydhaselton** [comment]
@qmfrederik and @janvorli: This may be a red herring but I noticed, when running strace on the 9809 commit build, that the assert failure was consistently thrown after an `ENOENT`

```
newfstatat(AT_FDCWD, "/data/data/com.termux/files/usr/bin/Product/Linux.arm64.Debug/en-US-POSIX", 0x7fe5458c28, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
```

Ignore if unrelated, if significant, see full strace here: http://pastebin.com/hRiA6dM7

**Comment by janvorli** [comment]
@cydhaselton what does it print to console if you run it without the strace? I can see from the strace that it was trying to write a string starting with `Assert failure(PID 26829`, but the strace doesn't show more of the string.

**Comment by cydhaselton** [comment]
@janvorli it prints the assert error that @qmfrederik posted earlier here: https://github.com/dotnet/coreclr/issues/9780#issuecomment-282597701

**Comment by janvorli** [comment]
I've examined the log in more detail. I can see that it cannot find System.Native.so. Maybe you've forgotten to add native components built in corefx repo?

**Comment by cydhaselton** [comment]
@janvorli I copied the corefx native components over as well; basically copied over the contents of ./Linux.arm64.Debug.

System.Native.so is in ./Linux.arm64.Debug/native

**Comment by janvorli** [comment]
It needs to be in the same folder as the managed assemblies, not in a standalone folder.

**Comment by janvorli** [comment]
This is true for everything. There should be no subfolders, everything from corefx and coreclr, managed and native should be in single folder.

**Comment by qmfrederik** [comment]
@janvorli @cydhaselton It has been my experience that on Android, dl_open does not always search the folder in which the executable is loaded for dynamic link libraries.

In my tests, I always needed to set `LD_LIBRARY_PATH` to the folder to which I deployed CoreCLR + CoreFX, so for example:

```
LD_LIBRARY_PATH=/data/local/tmp/coredroid/ /data/local/tmp/coredroid/corerun /data/local/tmp/coredroid/helloworld.dll
```

@cydhaselton my notes are here: https://github.com/qmfrederik/coredroid/blob/master/README.md

**Comment by janvorli** [comment]
@qmfrederik I have thought that paths in the LD_LIBRARY_PATH need to be separated by `:`, not by spaces.

**Comment by qmfrederik** [comment]
@janvorli You are right; so `LD_LIBRARY_PATH=...` sets the env variable only for the process which is about to be launched (corerun)

**Comment by cydhaselton** [comment]
@janvorli I have found that with coreclr on Android...at least within the Termux hosting application...the corerun search path overrides the LD_LIBRARY_PATH and that corerun will find libs in it's own directory over paths set by its environment vairables

@qmfrederik The builds produced by your scripts output files and subdirectories in `Linux.arm64.Debug`.  Is the recommendation to copy _all_ of the files in the generated subdirectories of `Linux.arm64.Debug` to the root Linux.arm64.Debug folder?

**Comment by qmfrederik** [comment]
@cydhaselton Nope, I don't think that's a recommendation - it's just what I do now to make things work :). I think there's a doc in this repo which describes which files you need to copy, just didn't get round to pinning down that list yet, so I took the lazy route.

**Comment by cydhaselton** [comment]
@qmfrederik Then I've copied the necessary files as I've seen the docs.  The assert error still occurs.

I've done some preliminary research on the PAL_VirtualUnwind error and what I've found seems to point to libunwind as being the culprit.  Does the version pulled by the scripts match the recommended version in the coreclr documentation?

**Comment by janvorli** [comment]
@cydhaselton So you now have the System.Native.so side by side with the libcoreclr.so? Can you share the strace log again?
